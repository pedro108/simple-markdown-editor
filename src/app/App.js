import { useCallback, useState } from "react";

import EditorButtons from "./components/EditorButtons";
import EditorContainer from "./components/EditorContainer";
import EditorPrimaryButton from "./components/EditorPrimaryButton";
import EditorWrapper from "./components/EditorWrapper";

function App() {
  const [markdown, setMarkdown] = useState("");
  const [result, setResult] = useState("");

  const handleChangeMarkdown = useCallback((e) => {
    setMarkdown(e.target.value);
  }, [setMarkdown]);

  const handleRenderButton = useCallback(() => {
    setResult(markdown);
  });

  return (
    <EditorContainer>
      <EditorWrapper>
        <textarea
          className="editor__editor"
          value={markdown}
          onChange={handleChangeMarkdown}
          data-testid="editor"
        />
        <div
          className="editor__result"
          data-testid="editor-result"
        >
          {result}
        </div>
      </EditorWrapper>
      <EditorButtons>
        <EditorPrimaryButton
          type="button"
          onClick={handleRenderButton}
          data-testid="render-button"
        >Render</EditorPrimaryButton>
      </EditorButtons>
    </EditorContainer>
  );
}

export default App;
