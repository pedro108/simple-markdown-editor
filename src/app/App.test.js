import { act, fireEvent, render } from '@testing-library/react';
import App from './App';
import markdownEngine from '../engine/markdown.engine';

describe('<App/>', () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it('Matches DOM snapshot', () => {
    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Parses the result from the markdown editor to the result box when Render button is clicked', () => {
    const { getByTestId } = render(<App />);
    const markdownValue = "#Title\n##Lorem ipsum dolor.";
    const expectedResult = markdownEngine.render(markdownValue);

    const editor = getByTestId("editor");
    const result = getByTestId("editor-result");
    const button = getByTestId("render-button");

    act(() => {
      fireEvent.change(editor, { target: { value: markdownValue } });
      fireEvent.click(button);
    });

    expect(result.innerHTML).toEqual(expectedResult);
  });

  it('Removes the render button if the live render toggle is checked', () => {
    const { getByTestId } = render(<App />);

    const liveRenderToggle = getByTestId("live-render-toggle");

    expect(liveRenderToggle.checked).toBe(false);
    expect(getByTestId("render-button")).toBeTruthy();

    act(() => {
      fireEvent.click(liveRenderToggle);
    });

    expect(liveRenderToggle.checked).toBe(true);
    expect(() => getByTestId("render-button")).toThrow();
  });

  it('Renders markdown automatically if live render toggled is checked', () => {
    const { getByTestId } = render(<App />);
    const markdownValue = "#Title\n##Lorem ipsum dolor.";
    const expectedResult = markdownEngine.render(markdownValue);

    const editor = getByTestId("editor");
    const result = getByTestId("editor-result");
    const liveRenderToggle = getByTestId("live-render-toggle");

    act(() => {
      fireEvent.click(liveRenderToggle);
    });

    act(() => {
      fireEvent.change(editor, { target: { value: markdownValue } });
    });

    expect(result.innerHTML).toEqual(expectedResult);
  });

  it('Reads from localStorage cache when rendering the editor for the first time', () => {
    const markdownValue = "#Title\n##Lorem ipsum dolor.";
    const expectedResult = markdownEngine.render(markdownValue);
    window.localStorage.setItem("markdown-cache", markdownValue);

    const { getByTestId } = render(<App />);

    const result = getByTestId("editor-result");

    expect(result.innerHTML).toEqual(expectedResult);    
  });

  it('Updates the localStorage cache when markdown is changed', () => {
    const { getByTestId } = render(<App />);
    const markdownValue = "#Title\n##Lorem ipsum dolor.";

    const editor = getByTestId("editor");
    const button = getByTestId("render-button");

    act(() => {
      fireEvent.change(editor, { target: { value: markdownValue } });
      fireEvent.click(button);
    });

    expect(window.localStorage.getItem("markdown-cache")).toEqual(markdownValue);
  });
});
