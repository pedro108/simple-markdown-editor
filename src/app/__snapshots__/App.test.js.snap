// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> Matches DOM snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-jCIOXC cClzMa"
  >
    <div
      class="sc-dBwcjI bjSTWf"
    >
      <textarea
        class="editor__editor"
        data-testid="editor"
      />
      <div
        class="editor__result"
        data-testid="editor-result"
      />
    </div>
    <div
      class="sc-hfDuwB kkTnSy"
    >
      <button
        class="sc-fBEEOr jhYYDd"
        data-testid="render-button"
        type="button"
      >
        Render
      </button>
    </div>
    <label>
      <input
        data-testid="live-render-toggle"
        type="checkbox"
      />
      Live render
    </label>
  </div>
</DocumentFragment>
`;
